File: src/core/validation/concurrentProducers.ts
import type { WorkflowModel } from '@core/model/WorkflowModel';
import type { ModelRegistry } from '@core/model/ModelRegistry';
import type { NodeId } from '@core/model/contracts/node';
import { ControlFlowGraph, type BranchPort, type IControlFlowGraph } from './controlFlowGraph';

/**
 * How many of a port's incoming links can carry a value in the **same run**.
 *
 * `workflow-gallery/64`. `maxConnections` on an input exists because a slot
 * holds one value: two producers writing it in one superstep is the ambiguity
 * the cap forbids. But it was counting *edges drawn*, and those are not the
 * same number. A router takes exactly one of its branches, so three branches
 * converging on one agent's `prompt` is three edges and one producer — while
 * two unrelated agents into that same prompt is two edges and two producers,
 * which is the case the cap is actually for (ticket 14).
 *
 * The distinction has to be structural, and it is: a `branch` output is
 * declared by its port as "one of several mutually exclusive ways out", and a
 * node whose branches are *not* exclusive says so (`INodeModel.branchesAreExclusive`
 * — a Router set to *run every match, in parallel* is exactly that node, and
 * discovering it is why this is asked of the node rather than read off the
 * port flag).
 *
 * Deliberately **conservative in one direction only**: where it cannot prove
 * two producers are exclusive it calls them concurrent, so the answer is never
 * smaller than the truth and the cap is never relaxed on a guess.
 */

/**
 * Which of a branching node's branches have to have been taken for `edge` to
 * carry a value, or `null` when this branching node does not decide that at
 * all — one row per branching node, one entry per edge.
 *
 * Two facts are needed per (branching node, edge) pair, and the whole of
 * `the-cost-of-one-more/21` is where they come from:
 *
 *  - **is the producer gated by this branching node at all** — it must be
 *    unreachable when the branching node is deleted. Without this, a node that
 *    can also be reached by a path going nowhere near the router would be
 *    treated as if the router gated it, and two genuinely concurrent producers
 *    would be called exclusive on the strength of a branch neither of them
 *    needs.
 *  - **which branches reach it**, which is what the constraint set names.
 *
 * Neither depends on the edge beyond its source, which is what `03` fixed.
 * Both were still answered by walking the whole document once per branching
 * node, which is what this pair of implementations fixes.
 */
type ConstraintRows = (Set<string> | null)[][];

/** The edge leaves the branching node itself: it carries exactly that branch. */
function branchLeavingOwner(branching: readonly BranchPort[], portId: string): Set<string> | null {
  const owner = branching[0]?.nodeId;
  if (owner === undefined) return null;
  const key = `${owner}#${portId}`;
  return branching.some((branch) => `${branch.nodeId}#${branch.portId}` === key)
    ? new Set([key])
    : null;
}

/**
 * The fast answer: one dominator pass, then set lookups.
 *
 * `null` when the index declines to answer for any producer — a control-flow
 * cycle, or a producer the roots cannot reach — in which case the caller falls
 * back to the walks, which are exact for those shapes too.
 *
 * **The invariant this rests on, named because `03` left it unnamed.** The
 * gating question is *"is the producer still reachable from the roots with the
 * owner deleted"*, and that is dominance: `owner` gates `producer` exactly when
 * it lies on every path to it. Dominance is a property of the **document**, not
 * of the branching node asking — so it is resolved once and read by every gate.
 *
 * The branch question keeps one thing from the walks and drops the rest. It
 * asks whether a branch target reaches the producer, and the walks ask that
 * *with the owner deleted*. The deletion is what stops a walk leaving by one
 * branch, looping back through the owner and arriving down another. It cannot
 * happen here: the index is `null` unless the graph the roots reach is
 * **acyclic**, and in an acyclic graph a successor of the owner has no path
 * back to it. So "reaches the producer" and "reaches the producer without the
 * owner" are the same set, and the producer's own ancestors — one walk per
 * *link on the port*, which `03` already paid for as its upstream filter —
 * answer it for every branching node at once.
 */
function constraintsByDominance(
  graph: IControlFlowGraph,
  edges: readonly { readonly source: { readonly nodeId: NodeId; readonly portId: string } }[],
): ConstraintRows | null {
  const gatedBy = new Map<NodeId, ReadonlySet<NodeId>>();
  const reaching = new Map<NodeId, ReadonlySet<NodeId>>();
  for (const producer of new Set(edges.map((edge) => edge.source.nodeId))) {
    const dominators = graph.dominatorsOf(producer);
    if (dominators === null) return null;
    gatedBy.set(producer, dominators);
    reaching.set(producer, graph.ancestorsOf(producer));
  }

  const rows: ConstraintRows = [];
  for (const branching of graph.branchingNodes) {
    const owner = branching[0]?.nodeId;
    if (owner === undefined) continue;
    // Resolved once per branching node rather than once per edge, which is
    // `03`'s rule applied to the two cheap questions this path has left.
    const branches = branching.map((branch) => ({
      key: `${branch.nodeId}#${branch.portId}`,
      targets: graph.targetsFrom(branch),
    }));
    const row = edges.map((edge) => {
      const producer = edge.source.nodeId;
      if (producer === owner) return branchLeavingOwner(branching, edge.source.portId);
      if (!(gatedBy.get(producer)?.has(owner) ?? false)) return null;
      const reaches = reaching.get(producer);
      const taken = new Set<string>();
      for (const branch of branches) {
        if (branch.targets.some((target) => reaches?.has(target) ?? false)) taken.add(branch.key);
      }
      return taken.size > 0 ? taken : null;
    });
    if (row.some((entry) => entry !== null)) rows.push(row);
  }
  return rows;
}

/**
 * The exact answer for a document the index declines: a walk of the whole
 * graph per branching node, plus one per branch it opens.
 *
 * This is `03`'s implementation, kept rather than replaced. It is quadratic in
 * the document and it is *correct on every shape*, including the cyclic ones
 * where reachability-with-a-node-deleted genuinely differs from reachability.
 * A control-flow cycle is drawable — `acyclicGraphRule` calls an escapable
 * loop a correct graph — so this path is reachable, and
 * `dominanceAgreesWithWalking.test.ts` runs the two against each other rather
 * than trusting the argument above.
 */
function constraintsByWalking(
  graph: IControlFlowGraph,
  edges: readonly { readonly source: { readonly nodeId: NodeId; readonly portId: string } }[],
): ConstraintRows {
  // A branching node can only gate a producer it can *reach*: a branch target
  // is downstream of its owner by construction, so the branching nodes
  // upstream of no producer answer `null` for every edge, and asking them
  // costs a walk of the whole document for a foregone conclusion. On the swap
  // gesture that is every router in the document.
  const upstream = new Set<NodeId>();
  for (const nodeId of new Set(edges.map((edge) => edge.source.nodeId))) {
    for (const ancestor of graph.ancestorsOf(nodeId)) upstream.add(ancestor);
  }

  const rows: ConstraintRows = [];
  for (const branching of graph.branchingNodes) {
    const owner = branching[0]?.nodeId;
    if (owner === undefined || !upstream.has(owner)) continue;
    const ungated = graph.reachable(graph.roots, { without: owner });
    const branches = branching.map((branch) => ({
      key: `${branch.nodeId}#${branch.portId}`,
      reached: graph.reachable(graph.targetsFrom(branch), { without: owner }),
    }));
    rows.push(
      edges.map((edge) => {
        const producer = edge.source.nodeId;
        if (producer === owner) return branchLeavingOwner(branching, edge.source.portId);
        if (ungated.has(producer)) return null;
        const taken = new Set<string>();
        for (const branch of branches) {
          if (branch.reached.has(producer)) taken.add(branch.key);
        }
        return taken.size > 0 ? taken : null;
      }),
    );
  }
  return rows;
}

/**
 * The largest number of these edges that can carry a value in one run.
 *
 * Exclusivity is a pairwise fact, and the largest set of *mutually* concurrent
 * edges is a maximum clique — so this returns the size of the largest connected
 * component of the "can coexist" graph instead, which is an upper bound on it.
 * Over-counting keeps a link refused or replaced that might have been allowed;
 * under-counting would let two values race into one slot. Only one of those is
 * survivable, and for the shapes this product actually draws — a router's
 * branches, a grader's verdicts — exclusivity is total and the bound is exact.
 */
export function concurrentProducerCountOn(
  graph: IControlFlowGraph,
  edges: readonly { readonly source: { readonly nodeId: NodeId; readonly portId: string } }[],
): number {
  if (edges.length < 2) return edges.length;
  if (graph.branchingNodes.length === 0) return edges.length;

  const constraints = constraintsByDominance(graph, edges) ?? constraintsByWalking(graph, edges);
  if (constraints.length === 0) return edges.length;

  const exclusive = (a: number, b: number): boolean =>
    constraints.some((perEdge) => {
      const left = perEdge[a];
      const right = perEdge[b];
      if (!left || !right) return false;
      return [...left].every((key) => !right.has(key));
    });

  // Connected components of "not proven exclusive".
  const component = edges.map((_, index) => index);
  const find = (index: number): number => {
    let root = index;
    while (component[root] !== root) root = component[root]!;
    return root;
  };
  for (let a = 0; a < edges.length; a += 1) {
    for (let b = a + 1; b < edges.length; b += 1) {
      if (exclusive(a, b)) continue;
      const rootA = find(a);
      const rootB = find(b);
      if (rootA !== rootB) component[rootB] = rootA;
    }
  }
  const sizes = new Map<number, number>();
  for (let index = 0; index < edges.length; index += 1) {
    const root = find(index);
    sizes.set(root, (sizes.get(root) ?? 0) + 1);
  }
  return Math.max(...sizes.values());
}

/**
 * The same question, asked of a document.
 *
 * Builds the control-flow skeleton and throws it away again: its lifetime is
 * this call, so there is nothing to invalidate. `capacityRule` asks this on
 * every pointer-move over a full port, and the model does not change during a
 * drag — a memo spanning the gesture would be faster still, and would need an
 * owner and an invalidation rule that no collaborator here has. That is a
 * different ticket, and only worth filing if this one is not enough.
 */
export function concurrentProducerCount(
  model: WorkflowModel,
  _registry: ModelRegistry,
  edges: readonly { readonly source: { readonly nodeId: NodeId; readonly portId: string } }[],
): number {
  if (edges.length < 2) return edges.length;
  return concurrentProducerCountOn(new ControlFlowGraph(model), edges);
}

File: backend/openstategraph/concurrent_producers.py
"""How many of a port's incoming links can carry a value in the **same run**.

`osg-agent-experience/43`. `maxConnections` on an input exists because a slot
holds one value, and two producers writing it in one superstep is the ambiguity
the cap forbids. The canvas has enforced that since `workflow-gallery/64`; a
document written by anything other than the canvas — an agent composing
`workflow.json` through the MCP door, which is how the try project was built —
met no such rule, and fifteen edges arrived on a one-slot input with every
checker green.

**Why this is not a raw edge count.** A router takes exactly one of its
branches, so three branches converging on one agent's `prompt` are three links
and one value. Two of this repository's own shipped examples do exactly that
(`support-triage`'s grader `candidate`, `chinook-assistant`'s `out1.result`), so
a checker that counted links would have refused documents that are correct and
that ship.

**This mirrors `src/core/validation/concurrentProducers.ts`, and says so.**
CLAUDE.md forbids hand-mirroring a *type* across the boundary and this is not
one — it is the same question asked of a document by a runtime that has no
`WorkflowModel` to ask. Two deliberate differences from that file, both
recorded rather than discovered later:

- **Only the walking implementation is here.** The TypeScript side also has a
  dominator index, because `capacityRule` runs on every pointer-move of a drag
  and paid 153 ms for the walks on a 516-node document
  (`the-cost-of-one-more/03`). `validate` runs once, per document, per command.
  A faster implementation whose agreement with the exact one needs its own test
  is a cost with nothing here to buy.
- **The conservatism runs the other way, and it must.** On the canvas, an
  answer that is too *large* refuses a link somebody could have drawn, and an
  answer that is too small lets two values race into one slot — so that side
  over-counts on a guess. Here an answer that is too large is a **false
  accusation against a valid document**, which is the one thing a checker may
  never do (`document_checks`' own `values_outside_their_options` records the
  same rule). So this side under-counts on a guess: where the canvas asks
  `INodeModel.branchesAreExclusive` and treats a Router in *run every match, in
  parallel* as a node whose branches really can race, this treats every
  `branch` output as exclusive, because the mode is node data and the
  catalogue publishes no such fact. The gap is real and named: a parallel
  router whose branches converge on one slot is a document this will not
  report.
"""

from __future__ import annotations

from collections.abc import Callable, Iterable, Mapping, Sequence
from dataclasses import dataclass, field
from typing import Any

#: Port types that wire *configuration*, not control flow — the same four
#: `controlFlowGraph.ts` names, for the same reason: walking them would report
#: an agent as reachable from its own tool palette, and a `feedback` edge
#: travels into the next superstep rather than this one.
NOT_CONTROL_FLOW = frozenset({"tool", "skill", "feedback", "worker"})

#: `(node id, port id) -> port type`, or `None` when this build cannot resolve
#: it — a package's own discovered tool, absent from a bare catalogue.
PortTypeResolver = Callable[[str, str], str | None]


@dataclass(frozen=True)
class _BranchPort:
    node_id: str
    port_id: str

    @property
    def key(self) -> str:
        return f"{self.node_id}#{self.port_id}"


@dataclass
class ControlFlow:
    """The document's control-flow skeleton, resolved once and asked many times."""

    successors: dict[str, list[str]] = field(default_factory=dict)
    branch_targets: dict[str, list[str]] = field(default_factory=dict)
    predecessors: dict[str, list[str]] = field(default_factory=dict)
    roots: tuple[str, ...] = ()
    branching_nodes: tuple[tuple[_BranchPort, ...], ...] = ()

    def targets_from(self, branch: _BranchPort) -> list[str]:
        return self.branch_targets.get(branch.key, [])

    def reachable(self, start: Iterable[str], *, without: str | None = None) -> set[str]:
        return _walk(self.successors, start, without)

    def ancestors_of(self, node_id: str) -> set[str]:
        return _walk(self.predecessors, [node_id], None)


def _walk(
    adjacency: Mapping[str, list[str]],
    start: Iterable[str],
    without: str | None,
) -> set[str]:
    seen: set[str] = set()
    queue: list[str] = []
    for node_id in start:
        if node_id == without or node_id in seen:
            continue
        seen.add(node_id)
        queue.append(node_id)
    head = 0
    while head < len(queue):
        current = queue[head]
        head += 1
        for nxt in adjacency.get(current, ()):
            if nxt == without or nxt in seen:
                continue
            seen.add(nxt)
            queue.append(nxt)
    return seen


def build_control_flow(
    node_ids: Sequence[str],
    edges: Sequence[Mapping[str, Any]],
    port_type: PortTypeResolver,
    branch_ports_of: Callable[[str], Sequence[str]],
) -> ControlFlow:
    """One pass over the document.

    `branch_ports_of` answers with a node's `branch` out-port ids — the
    catalogue's own `branch` flag (artifact schema 5), resolved by the caller
    because only it knows how a node's configured ports are named.

    The roots are computed on the whole graph, before any branching node is
    hypothetically deleted: recomputing them afterwards would promote every
    node the deletion orphaned into a root, and then nothing would be gated by
    anything.
    """
    flow = ControlFlow()
    fed: set[str] = set()
    for edge in edges:
        source = edge.get("source")
        target = edge.get("target")
        if not isinstance(source, dict) or not isinstance(target, dict):
            continue
        src_id = str(source.get("nodeId") or "")
        dst_id = str(target.get("nodeId") or "")
        if not src_id or not dst_id:
            continue
        from_type = port_type(src_id, str(source.get("portId") or ""))
        to_type = port_type(dst_id, str(target.get("portId") or ""))
        # Both ends are asked, and either one saying no is enough — the
        # consumer is the end that is always a catalogue node, and `tools` is
        # `tool` whoever is plugged into it.
        if from_type is None or to_type is None:
            continue
        if from_type in NOT_CONTROL_FLOW or to_type in NOT_CONTROL_FLOW:
            continue
        flow.successors.setdefault(src_id, []).append(dst_id)
        flow.predecessors.setdefault(dst_id, []).append(src_id)
        flow.branch_targets.setdefault(
            f"{src_id}#{source.get('portId') or ''}", []
        ).append(dst_id)
        fed.add(dst_id)

    flow.roots = tuple(node_id for node_id in node_ids if node_id not in fed)
    branching: list[tuple[_BranchPort, ...]] = []
    for node_id in node_ids:
        ports = [_BranchPort(node_id, port_id) for port_id in branch_ports_of(node_id)]
        # Fewer than two ways out is not a choice between them.
        if len(ports) >= 2:
            branching.append(tuple(ports))
    flow.branching_nodes = tuple(branching)
    return flow


def _branch_leaving_owner(
    branching: Sequence[_BranchPort], port_id: str
) -> set[str] | None:
    """The edge leaves the branching node itself: it carries exactly that branch."""
    owner = branching[0].node_id
    key = f"{owner}#{port_id}"
    return {key} if any(branch.key == key for branch in branching) else None


def concurrent_producer_count(
    flow: ControlFlow, edges: Sequence[Mapping[str, Any]]
) -> int:
    """The largest number of `edges` that can carry a value in one run.

    Exclusivity is a pairwise fact and the largest set of *mutually* concurrent
    edges is a maximum clique, so this returns the size of the largest
    connected component of the "can coexist" graph — an upper bound on it. For
    the shapes this product draws (a router's branches, a grader's verdicts)
    exclusivity is total and the bound is exact.
    """
    if len(edges) < 2 or not flow.branching_nodes:
        return len(edges)

    producers = [str((edge.get("source") or {}).get("nodeId") or "") for edge in edges]
    ports = [str((edge.get("source") or {}).get("portId") or "") for edge in edges]

    # A branching node can only gate a producer it can reach, so the ones
    # upstream of none of them answer `None` for every edge.
    upstream: set[str] = set()
    for producer in set(producers):
        upstream |= flow.ancestors_of(producer)

    constraints: list[list[set[str] | None]] = []
    for branching in flow.branching_nodes:
        owner = branching[0].node_id
        if owner not in upstream:
            continue
        ungated = flow.reachable(flow.roots, without=owner)
        reached = {
            branch.key: flow.reachable(flow.targets_from(branch), without=owner)
            for branch in branching
        }
        row: list[set[str] | None] = []
        for producer, port_id in zip(producers, ports, strict=True):
            if producer == owner:
                row.append(_branch_leaving_owner(branching, port_id))
                continue
            if producer in ungated:
                row.append(None)
                continue
            taken = {key for key, seen in reached.items() if producer in seen}
            row.append(taken or None)
        if any(entry is not None for entry in row):
            constraints.append(row)

    if not constraints:
        return len(edges)

    def exclusive(a: int, b: int) -> bool:
        for row in constraints:
            left, right = row[a], row[b]
            if left is None or right is None:
                continue
            if left.isdisjoint(right):
                return True
        return False

    parent = list(range(len(edges)))

    def find(index: int) -> int:
        while parent[index] != index:
            index = parent[index]
        return index

    for a in range(len(edges)):
        for b in range(a + 1, len(edges)):
            if exclusive(a, b):
                continue
            root_a, root_b = find(a), find(b)
            if root_a != root_b:
                parent[root_b] = root_a

    sizes: dict[int, int] = {}
    for index in range(len(edges)):
        root = find(index)
        sizes[root] = sizes.get(root, 0) + 1
    return max(sizes.values())


__all__ = [
    "NOT_CONTROL_FLOW",
    "ControlFlow",
    "build_control_flow",
    "concurrent_producer_count",
]
